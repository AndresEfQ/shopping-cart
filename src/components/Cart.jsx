import { useContext } from "react";
import { CartContext } from "../context/CartContext";
import styled from "styled-components";
import { FaTrashAlt } from "react-icons/fa";

export default function Cart({toggleCart}) {

  const {cart, setCart} = useContext(CartContext);

  const subtotal = cart.items.reduce((current, item) => current + (item.price * item.number), 0)

  const removeFromCart = (e) => {

    setCart(prevCart => {
      const itemToRemove = prevCart.items.find(item => item.name === e.target.dataset.name)
      console.log(itemToRemove);
      if (itemToRemove.number === 1) {
        return ({
          items: prevCart.items.filter(item => item.name !== itemToRemove.name)
        })
      } else {
        return ({
          items: prevCart.items.map(item => {
            if (item.name === itemToRemove.name) {
              return {...item, number: item.number - 1}
            } else {
              return item;
            }
          })
        })
      }
    })
  }

  return (
    <CartDiv onClick={toggleCart}>
      {cart.items.length ? 
        cart.items.map(item => 
          <CartItem onClick={(e) => e.stopPropagation()} key={item.name}>
            <img src={item.img} alt={item.name} />
            <InfoDiv>
              <h3>{item.name}</h3>
              <ul>
                <li>
                  <span>Cards Ammount:</span>
                  <span data-testid="cards-number">{item.number}</span>
                </li>
                <li>
                  <span>Total:</span>
                  <span data-testid="cards-price">
                    {new Intl.NumberFormat('en-US', {
                      style: 'currency', 
                      currency: 'USD'
                      }).format(item.price * item.number)
                    }
                  </span>
                </li>
              </ul>
            </InfoDiv>
            <button 
              onClick={removeFromCart} 
              data-name={item.name} 
              data-testid="remove-from-cart"
            >
              <FaTrashAlt color="var(--grey)" />
            </button>
          </CartItem>) : 
        <CartItem onClick={(e) => e.stopPropagation()}>
          <em>There are no items in the cart</em>
        </CartItem>
      }
      <hr />
      <CartItem onClick={(e) => e.stopPropagation()}>
        <span>Subtotal</span>
        <span data-testid="subtotal">
          {new Intl.NumberFormat('en-US', {
            style: 'currency', 
            currency: 'USD'
            }).format(subtotal)
          }
        </span>
      </CartItem>
    </CartDiv>
  )
}

const CartDiv = styled.div`
  position: absolute;
  right: 0;
  width: 100%;
  height: 88vh;
  background-color: var(--secondary-op90);
  padding: 1rem;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  gap: 1rem;

  & hr {
    width: 20vw;

    @media only screen and (max-width: 490px) {
      width: 100%;
    }
  }
`;

const CartItem = styled.div`
  width: 20vw;
  background-color: var(--op80);
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 0.5rem 2rem;

  @media only screen and (max-width: 490px) {
    width: 100%;
    padding: 0.5rem;
  }

  & img {
    width: 15%;

    @media only screen and (max-width: 490px) {
      width: 25%;
    }
  }

  & svg {
    pointer-events: none;

    @media only screen and (max-width: 490px) {
      transform: scale(2);
    }
  }

  & button {
    border: none;
    background: none;
    cursor: pointer;

    @media only screen and (max-width: 490px) {
      width: 2rem;
      height: 2rem;
    }
  }

  & button:active {
    transform: scale(0.95);
  }
`;

const InfoDiv = styled.div`
  width: 10vw;

  @media only screen and (max-width: 490px) {
    width: 50vw;
  }

  & h3 {
    border-bottom: 1px solid var(--grey);
  }
  
  & ul {
    list-style: none;
  }
  
  & li {
    display: flex;
    justify-content: space-between;
  }
`;

