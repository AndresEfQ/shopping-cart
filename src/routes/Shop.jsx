import { useEffect, useState } from "react";
import styled from "styled-components";
import SimpleBar from "simplebar-react";
import "simplebar-react/dist/simplebar.min.css"
import { FaSearch } from "react-icons/fa";
import DotLoader from "react-spinners/DotLoader";
import CardsList from "../components/CardsList";
import backupData from "../assets/cache";

export default function Shop({windowWidth}) {

  const [apiResponse, setApiResponse] = useState();
  const [sets, setSets] = useState();
  const [cards, setCards] = useState(backupData.cards);
  const [isLoadingCards, setIsLoadingCards] = useState(false);
  const [searchInput, setSearchInput] = useState("");

  useEffect(() => {

    fetch("https://api.magicthegathering.io/v1/sets")
    .then(response => response.json())
    .then(response => {
      setApiResponse(response.sets);
      setSets(response.sets);
    })
    .catch(err => console.log(err));

  }, []);

  const searchSets = (query, array) => {
    return array.filter(element => element.name.toLowerCase().indexOf(query.toLowerCase()) !== -1);
  }
  
  const handleSearchInputChange = (e) => {
    setSearchInput(e.target.value);
    setSets(searchSets(e.target.value, apiResponse));
  }

  const handleSelectSet = (e) => {
    e.preventDefault();
    setIsLoadingCards(true);
    const set = e.target.id || e.target[0].value
    console.log(e.target[0].value)
    fetch(`https://api.magicthegathering.io/v1/cards?set=${set}&random=true&pageSize=12`)
    .then(response => response.json())
    .then(response => {
      setCards(response.cards);
      setIsLoadingCards(false);
      console.log(response.cards)
    })
    .catch(err => console.log(err));
  }

  return (
    <ShopDiv>
      {(windowWidth > 490) ? 
        <div>
          <SearchBar
            type="search"
            placeholder="Search Sets"
            onChange={handleSearchInputChange}
            value={searchInput}
          />
          <FaSearch color="var(--grey)"/>
          <SideBar>
            <SimpleBar style={{maxHeight: "80vh", width: "15vw"}}>
              {sets
                ? sets.map(set => {
                  return (
                    <SetSelector
                      key={set.code}
                      id={set.code}
                      onClick={handleSelectSet}>{set.name}
                    </SetSelector>
                  )})
                : <span>loading</span>}
            </SimpleBar>
          </SideBar>
        </div> : 
        <form onSubmit={handleSelectSet}>
          <label htmlFor="sets">Choose a set:</label>
          <input list="setsList" id="sets" name="sets" />
          <datalist id="setsList">
            {sets && sets.map(set => {
              return (
                <option
                  key={set.code}
                  data-code={set.code}
                  value={set.code}
                >{set.name}</option>
              )})}
          </datalist>
          <button>OK</button>
        </form>
      }
      {cards ? <CardsList cards={cards} /> : <Shade><DotLoader color="var(--main)" /></Shade>}
      {isLoadingCards && <Shade><DotLoader color="var(--main)" /></Shade>}            
    </ShopDiv>
  )
}

const ShopDiv = styled.div`

  position: relative;
  display: flex;

  @media only screen and (max-width: 490px) {
    flex-direction: column;
    width: 100%
  }

  & > div > svg {
    position: absolute;
    top: 2.6vh;
    left: 2rem;
  }

  & > input {
    width: 60%;
  }
`

const SideBar = styled.aside`
  height: 80vh;
  width: 15vw;
  margin-left: 1rem;
  margin-top: 2vh;
  position: relative;
  display: flex;
  flex-direction: column;
  overflow: auto;
  background-color: var(--secondary-op60);
  scrollbar-width: none;
  -ms-overflow-style: none;
`;

const SearchBar = styled.input`
  margin-left: 1.5rem;
  margin-top: 2vh;
  padding: 0.2rem 1.8rem;
  width: 13vw;
`;

const SetSelector = styled.button`
  width: 14vw;
  margin: 0.5vh;
  padding: 0.4rem 0.8rem;
  overflow-x: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  text-align: left;
  cursor: pointer;
`;

const Shade = styled.div`
  height: 88vh;
  width: 100vw;
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  background: var(--op80);
  z-index: 3;
`;
