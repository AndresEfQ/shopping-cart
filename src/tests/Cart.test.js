import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Shop from "../routes/Shop";
import Cart from "../components/Cart";
import CartContextProvider, { CartContext } from "../context/CartContext"

const customCart = {
  cart: {
    items: [
      {
        name: "test card 1",
        img: "test img 1",
        number: 2,
        price: 0.5,
      },
      {
        name: "test card 2",
        img: "test img 2",
        number: 2,
        price: 1,
      }
    ]
  },
  setCart: () => {},
}

describe("Cart component", () => {

  test("Cart renders correctly", () => {
    const { container } = render(<CartContextProvider><Cart /></CartContextProvider>);

    expect(container).toMatchSnapshot();
  });

  test("Renders cart items", () => {
    render(<CartContext.Provider value={customCart}><Cart /></CartContext.Provider>);
    expect(screen.getByRole("heading", {name: "test card 1"})).toBeInTheDocument();
  });

  test("Renders cards number correctly", () => {
    render(<CartContext.Provider value={customCart}><Cart /></CartContext.Provider>);
    expect(screen.getAllByTestId("cards-number")[0]).toHaveTextContent("2");
  });
  
  test("Calculates cards price correctly", () => {
    render(<CartContext.Provider value={customCart}><Cart /></CartContext.Provider>);
    expect(screen.getAllByTestId("cards-price")[0]).toHaveTextContent("1");
  });

  test("Calculates subtotal price correctly", () => {
    render(<CartContext.Provider value={customCart}><Cart /></CartContext.Provider>);
    expect(screen.getByTestId("subtotal")).toHaveTextContent("3");
  });

});