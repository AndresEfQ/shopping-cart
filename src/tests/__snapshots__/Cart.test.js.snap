// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component Cart renders correctly 1`] = `
<div>
  <div
    class="sc-fsQiph dRhvcl"
  >
    <div
      class="sc-qRumB dheAZg"
    >
      <em>
        There are no items in the cart
      </em>
    </div>
    <hr />
    <div
      class="sc-qRumB dheAZg"
    >
      <span>
        Subtotal
      </span>
      <span
        data-testid="subtotal"
      >
        $0.00
      </span>
    </div>
  </div>
</div>
`;
